examples/topic_mirror.rs

```diff
@@ -53,34 +52,61 @@ fn create_consumer(brokers: &str, group_id: &str, topic: &str) -> LoggingConsume
         .create_with_context::<LoggingContext, LoggingConsumer>(LoggingContext)
         .expect("Consumer creation failed");
 
-    consumer.subscribe(&[topic]).expect("Can't subscribe to specified topic");
+    consumer
+        .subscribe(&[topic])
+        .expect("Can't subscribe to specified topic");
 
     consumer
 }
 
 struct MirrorProducerContext {
-    source_consumer: Arc<LoggingConsumer>
+    source_consumer: Arc<LoggingConsumer>,
 }
 
 impl ClientContext for MirrorProducerContext {}
 impl ProducerContext for MirrorProducerContext {
-    type DeliveryOpaque = TopicPartitionList;
+    type DeliveryOpaque = ();
 
-    fn delivery(&self, delivery_result: &Result<BorrowedMessage, (KafkaError, BorrowedMessage)>, delivery_opaque: TopicPartitionList) {
+    fn delivery(&self, delivery_result: &Result<BorrowedMessage, (KafkaError, BorrowedMessage)>, _delivery_opaque: ()) {
         match delivery_result {
             Ok(m) => {
-                info!("Delivered copy of {:?} to {}", delivery_opaque, m.topic());
-                self.source_consumer.store_offset_list(&delivery_opaque);
-            },
-            Err((e, m)) => error!("Failed to deliver to {}: {:?}", m.topic(), e)
+                debug!("Delivered copy of {:?} to {}", m.payload().unwrap(), m.topic());
+                // Question(blt) -- This call to `store_offset` will eventually
+                // cause offsets to be committed out of order. That will cause
+                // data loss. Should, then, the author cobble together a MinHeap
+                // (or similar) approach to commit only when all the suitable
+                // offsets have come through -- implying storage here -- or rely
+                // on rdkafka to take care of this?
```

Comment:
`librdkafka` doesn't provide a way to commit in order, so yes, we should implement a different way to keep track of them and only periodically commit the lowest one per partition.

Reply:
Here, by rdkafka, I meant this crate, not `librdkafka`.

Reply:
Whether rdkafka would be the right place for this functionality or not, I'm still a touch confused about how to pull this off, mostly down to lifetime management. 

In this example, the `TopicPartitionList` will always have one originating message associated with it. My initial idea was to `TopicPartitionList::elements` to retrieve the single `TopicPartitionListElem` and store that in an `BinaryHeap`. There's no `Ord` impl for `TopicPartitionListElem` yet—resolvable—but the key difficulty is the lack of Sync on `*mut rdkafka_sys::bindings::macos_64::rd_kafka_topic_partition_list_s`. My latest commit has the full code changes to reproduce, but here's a sample of the compilation error:

```
error[E0277]: `*mut rdkafka_sys::bindings::macos_64::rd_kafka_topic_partition_list_s` cannot be shared between threads safely
  --> examples/topic_mirror.rs:70:10
   |
70 | impl<'a> ProducerContext for MirrorProducerContext<'a> {
   |          ^^^^^^^^^^^^^^^ `*mut rdkafka_sys::bindings::macos_64::rd_kafka_topic_partition_list_s` cannot be shared between threads safely
   |
   = help: within `rdkafka::topic_partition_list::TopicPartitionListElem<'a>`, the trait `std::marker::Sync` is not implemented for `*mut rdkafka_sys::bindings::macos_64::rd_kafka_topic_partition_list_s`
   = note: required because it appears within the type `rdkafka::TopicPartitionList`
   = note: required because it appears within the type `&'a rdkafka::TopicPartitionList`
   = note: required because it appears within the type `rdkafka::topic_partition_list::TopicPartitionListElem<'a>`
   = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<rdkafka::topic_partition_list::TopicPartitionListElem<'a>>`
   = note: required because it appears within the type `alloc::raw_vec::RawVec<rdkafka::topic_partition_list::TopicPartitionListElem<'a>>`
   = note: required because it appears within the type `std::vec::Vec<rdkafka::topic_partition_list::TopicPartitionListElem<'a>>`
   = note: required because it appears within the type `std::collections::BinaryHeap<rdkafka::topic_partition_list::TopicPartitionListElem<'a>>`
   = note: required because it appears within the type `MirrorProducerContext<'a>`
```